### PR TITLE
[proposal] Improve configuration by using env variables correctly

### PR DESCRIPTION
**Pull request by vrde**
Hey, this is a proposal to improve the configuration framework. This should make the life of docker users easier (see #36, #49).

You can find most of the information [here](https://github.com/bigchaindb/bigchaindb/blob/improve-config/docs/source/bigchaindb-cli.md).

### CODE REVIEW

File: docs/source/bigchaindb-cli.md [hunk 1]
```diff
@@ -20,3 +21,65 @@ There's another command named `bigchaindb-benchmark`. It's used to run benchmark
...
 ```
+
+Note that you can always start `bigchaindb` using a different config file using the `-c` option.
+For more information check the help with `bigchaindb -h`.
+
+
+# Precedence in reading configuration values
+
+Note that there is a precedence in reading configuration values:
+ - (**not yet**) command line;
```
Comment (ttmc) [inline docs/source/bigchaindb-cli.md]: Maybe remove this line (about config values set from the command line) for now, until we actually implement that (if ever).

File: docs/source/bigchaindb-cli.md [hunk 2]
```diff
@@ -20,3 +21,65 @@ There's another command named `bigchaindb-benchmark`. It's used to run benchmark
...
+{...
+    "database": {"host": "localhost",
+                 "port": 28015}
+...}
+```
+
+while your local file `local.json` contains:
+```
+{...
+    "database": {"host": "whatever"}
```
Comment (ttmc) [inline docs/source/bigchaindb-cli.md]: Could you give a more concrete example than "whatever"? Maybe show a typical (fake) public DNS name?

File: docs/source/bigchaindb-cli.md [hunk 3]
```diff
@@ -20,3 +21,65 @@ There's another command named `bigchaindb-benchmark`. It's used to run benchmark
...
+```
+
+# List of env variables
+
+```
+- BIGCHAINDB_DATABASE_HOST
+- BIGCHAINDB_DATABASE_PORT
+- BIGCHAINDB_DATABASE_NAME
+- BIGCHAINDB_KEYPAIR_PUBLIC
+- BIGCHAINDB_KEYPAIR_PRIVATE
```
Comment (ttmc) [inline docs/source/bigchaindb-cli.md]: Need to add environment variables for the keyring (list of public keys of federation nodes).

Maybe also have an example of setting those values (e.g. in a `.bashrc` file), with actual values shown (i.e. a template)?

Comment (vrde) [inline docs/source/bigchaindb-cli.md]: > Need to add environment variables for the keyring (list of public keys of federation nodes). Maybe BIGCHAINDB_KEYRING?

done!

> Maybe also have an example of setting those values (e.g. in a .bashrc file), with typical/realistic values shown (i.e. a template)?

I think it's better if we create a new doc page describing all the config values, what do you think? we can do it after separating the client code from the server code.

File: bigchaindb/commands/bigchain.py [hunk 4]
```diff
@@ -88,7 +88,7 @@ def run_drop(args):
 
 def run_start(args):
     """Start the processes to run the node"""
-    run_configure(args, skip_if_exists=True)
+    # run_configure(args, skip_if_exists=True)
```
Comment (vrde) [inline bigchaindb/commands/bigchain.py]: @ttmc: please note that I disabled the "autoconfiguration" feature. We might need to update the docs.

File: docs/source/bigchaindb-cli.md (hunk 3 above)
Comment (ttmc) [inline docs/source/bigchaindb-cli.md]: Yes a new doc page would be nice, but you're right: that can wait until after the separation of the client and server code (and configs).

File: bigchaindb/config_utils.py [hunk 5]
```diff
@@ -57,13 +86,75 @@ def file_config(filename=None):
...
+        return os.environ.get(var_name, value)
 
-def dict_config(newconfig):
+    return map_leafs(load_from_env, config)
+
+
+def update_types(config):
+    """Return a new configuration where all the values types
+    are aligned with the ones in the default configuration"""
+
```
Comment (diminator) [inline bigchaindb/config_utils.py]: this is somewhat unclear to me from the description

it is a quite complex function, maybe give an example on how it works?

File: bigchaindb/config_utils.py [hunk 6]
```diff
@@ -57,13 +86,75 @@ def file_config(filename=None):
...
+        # Coerce a value to the `current` type.
+        try:
+            # First we try to apply current to the value, since it
+            # might be a function
+            return current(value)
+        except TypeError:
+            # Then we check if current is a list AND if the value
+            # is a string.
+            if isinstance(current, list) and isinstance(value, str):
+                # If so, we use the colon as the separator
```
Comment (diminator) [inline bigchaindb/config_utils.py]: why both checks?

File: bigchaindb/config_utils.py [hunk 7]
```diff
@@ -23,11 +23,40 @@
...
+        return mapping
+
+    return _inner(copy.deepcopy(mapping))
+
+
+def get_casted_value(original, to_convert):
+    try:
+        return original(to_convert)
+    except TypeError:
+        return type(original)(to_convert)
```
Comment (diminator) [inline bigchaindb/config_utils.py]: this is weird?

when do you get typeerrors?
what about valueerrors?

**Comment by rhsimplex** [comment]
```
 $ bigchaindb -c config.json start
Cannot find config file `/home/ryan/.bigchaindb`.
INFO:bigchaindb.config_utils:Configuration loaded from `config.json`
INFO:bigchaindb.commands.bigchain:Start bigchaindb main process
INFO:bigchaindb.processes:Initializing BigchainDB...
INFO:bigchaindb.config_utils:System already configured, skipping autoconfiguration
Traceback (most recent call last):
  File "/home/ryan/bigchaindb-testenv/bigchaindb/env/bin/bigchaindb", line 9, in <module>
    load_entry_point('BigchainDB==0.1.4', 'console_scripts', 'bigchaindb')()
  File "/home/ryan/bigchaindb-testenv/bigchaindb/env/lib/python3.4/site-packages/BigchainDB-0.1.4-py3.4.egg/bigchaindb/commands/bigchain.py", line 130, in main
    start(parser, globals())
  File "/home/ryan/bigchaindb-testenv/bigchaindb/env/lib/python3.4/site-packages/BigchainDB-0.1.4-py3.4.egg/bigchaindb/commands/utils.py", line 41, in start
    func(args)
  File "/home/ryan/bigchaindb-testenv/bigchaindb/env/lib/python3.4/site-packages/BigchainDB-0.1.4-py3.4.egg/bigchaindb/commands/bigchain.py", line 99, in run_start
    processes.start()
  File "/home/ryan/bigchaindb-testenv/bigchaindb/env/lib/python3.4/site-packages/BigchainDB-0.1.4-py3.4.egg/bigchaindb/processes.py", line 71, in start
    webapi = server.create_app()
  File "/home/ryan/bigchaindb-testenv/bigchaindb/env/lib/python3.4/site-packages/BigchainDB-0.1.4-py3.4.egg/bigchaindb/web/server.py", line 18, in create_app
    app.config['bigchain'] = Bigchain()
  File "/home/ryan/bigchaindb-testenv/bigchaindb/env/lib/python3.4/site-packages/BigchainDB-0.1.4-py3.4.egg/bigchaindb/core.py", line 59, in __init__
    raise exceptions.KeypairNotFoundException()
bigchaindb.exceptions.KeypairNotFoundException
```

```
$ cat config.json 
{
    "api_endpoint": "http://localhost:8008/api/v1",
    "database": {
        "host": "localhost",
        "name": "bigchain",
        "port": 28015
    },
    "keypair": {
        "private": "9PkLfHbzXnSSNnb1sSBL73C2MydzKLs5fAHoA4Q7otrG",
        "public": "26y9EuyGP44JXxqcvF8GbCJGqkiqFXddZzxVjLU3rWbHp"
    },
    "keyring": [
        "cZicZETBjq9vmetQX3QZJKH3h9NVieR8UGHsFJawkFz8",
        "bjkhDuc48iKoiChdtzyWDifAoU3F3EtYG876pjiEqjHt",
        "28PRHccTciwk82erw3bCFaDJARHYRkp3dnKDKcraitpir",
        "282jrCd3ZTX5rFKryzkdWNFmpVRJb68pWaNjYFD95MNp6",
        "xHxXRQzjxNDt1wSQQPksz98cprFJswhwZ4VXyKW2SHnP",
        "tJNvqs3Y3SEscqmbBTAkDcdgDQEdVFfhX6uDrsUBG4So",
        "26JBwKLQJEhqFqoNbJR3sfKSH4Aubx18ahtCcreMpayrE",
        "29e9kFnXeNBGSyfgYCCjzic7Tmkfasc9cdo2ZDw7gaxw2",
        "jvugmTr93MxbxRjkqfpihgqJi5GdXXCmRWjcGX6PL1No",
        "toaSC1jH4UdQYmQSrtmRTEZxjnaCU27idD4Zk63GKwGd",
        "ceDrgyRzTTPq42KrnJHXKcuaJw63hxKo65BAzf7sA7it",
        "tXAsTtqVLd7sqWSmCRMuacz4bZjYdVk9iCnsNWp5A5S2",
        "nJN74KZAq7e3zvC5gZkDREwfchrZxJd9DaYbHhdWPo5p",
        "27ogSfvWt19ry8SaWDAHY1xam2ZLJA2VyTddkgsnzuJFR",
        "x9QLnZ72oiZgYjSwSFuLuYJ2Vczg4XLmgRLcp4UKToqu",
        "ojLhidoD8hpW8ZXkccEKbsgDgzi2A5djNs2b69setmQA",
        "dPhYfMGxZyVtjgYojtjuGAd62oTVqABVtmsRF1ffkwyv",
        "oQd9xqF4SPEtExyed4QyErNnN6kP86Vh2R1ay3b7cH1F",
        "2A3nJmeMECvWAHpMUpL7ByqkLwsSywDAXgwuRpgLui3HT"
    ],
    "statsd": {
        "host": "localhost",
        "port": 8125,
        "rate": 0.01
    }
}
```

**Comment by vrde** [comment]
Thanks @rhsimplex for finding the bug. Should work now!

**Comment by rhsimplex** [comment]
Looks good to me! Thanks @vrde

**Comment by vrde** [comment]
Since we are using Gunicorn as app server (see #155) I need to update some of the code here.

**Comment by vrde** [comment]
@ttmc I feel the need of another section in the docs on how to configure the system. I think it's better if I work on it in this PR, what do you think? AFAIK there is no specific place right now in the docs about configuring the system.

**Comment by ttmc** [comment]
@vrde Okay, good idea. I will watch for a commit to this branch that adds a new section to the documentation. Maybe call it `configuration.md`?

**Comment by vrde** [comment]
> Maybe call it configuration.md?

Yup, working on it right nao.

File: bigchaindb/config_utils.py (hunk 5 above)
Comment (vrde) [inline bigchaindb/config_utils.py]: Does [this](https://github.com/bigchaindb/bigchaindb/pull/153/files#diff-377d993ccb19fa21dd579a661ff1c411R96) help? I added an explicit test.

File: bigchaindb/config_utils.py (hunk 7 above)
Comment (vrde) [inline bigchaindb/config_utils.py]: it's useless actually (the logic has been moved to `update_types`).

**Comment by ttmc** [comment]
@vrde I made some edits to `configuration.md` and related documents, and added that file to the Table of Contents (`index.rst`).

**Comment by vrde** [comment]
> and added that file to the Table of Contents

image

good point!

**Pull request merged**